src/gui/iconutils.cpp

```diff
@@ -1,52 +1,140 @@
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
 #include "iconutils.h"
 
 #include <theme.h>
 
 #include <QFile>
+#include <QLoggingCategory>
 #include <QPainter>
 #include <QPixmapCache>
 #include <QSvgRenderer>
 
+namespace {
+QString findSvgFilePath(const QString &fileName, const QStringList &possibleColors)
+{
+    const QString baseSvgNoColor{QString{OCC::Theme::themePrefix} + fileName};
+    if (QFile::exists(baseSvgNoColor)) {
+        return baseSvgNoColor;
+    }
+
+    for (const auto &color : possibleColors) {
+        const QString baseSVG{QString{OCC::Theme::themePrefix} + color + QLatin1Char('/') + fileName};
+
+        if (QFile::exists(baseSVG)) {
+            return baseSVG;
+        }
+    }
+
+    return {};
+}
+}
+
 namespace OCC {
 namespace Ui {
-namespace IconUtils {
-QPixmap pixmapForBackground(const QString &fileName, const QColor &backgroundColor)
-{
-    Q_ASSERT(!fileName.isEmpty());
+    Q_LOGGING_CATEGORY(lcIconUtils, "nextcloud.gui.iconutils", QtInfoMsg)
+    namespace IconUtils {
+        QPixmap pixmapForBackground(const QString &fileName, const QColor &backgroundColor)
+        {
+            Q_ASSERT(!fileName.isEmpty());
+
+            const auto pixmapColor = backgroundColor.isValid()
+                && !Theme::isDarkColor(backgroundColor)
+                ? QColorConstants::Svg::black
+                : QColorConstants::Svg::white;
+
+            return createSvgPixmapWithCustomColor(fileName, pixmapColor);
+        }
+
+        QPixmap createSvgPixmapWithCustomColor(const QString &fileName, const QColor &customColor, const QSize &size)
+        {
+            Q_ASSERT(!fileName.isEmpty());
+            Q_ASSERT(customColor.isValid());
+
+            if (fileName.isEmpty()) {
+                qWarning(lcIconUtils) << "fileName is empty";
+            }
+
+            if (!customColor.isValid()) {
+                qWarning(lcIconUtils) << "customColor is invalid";
+            }
+
+            const auto customColorName = customColor.name();
+
+            const QString cacheKey = fileName + QLatin1Char(',') + customColorName;
 
-    // some icons are present in white or black only, so, we need to check both when needed
-    const auto iconBaseColors = QStringList({ QStringLiteral("black"), QStringLiteral("white") });
+            QPixmap cachedPixmap;
 
-    const QString pixmapColor = backgroundColor.isValid() && !Theme::isDarkColor(backgroundColor) ? "black" : "white";
+            // check for existing QPixmap in cache
+            if (QPixmapCache::find(cacheKey, &cachedPixmap)) {
+                return cachedPixmap;
+            }
+
+            // some icons are present in white or black only, so, we need to check both when needed
+            const auto iconBaseColors = QStringList{QStringLiteral("black"), QStringLiteral("white")};
```

Comment:
Could we assure that we only have black or white svg images or at least make sure that we have from every svg image both versions?  Or is that too much work? That would simplify the code, and in the future, it would simplify everything a lot if we only have one color of every svg image.

Reply:
@FlexW That means a lot of work replacing all the `svgs`, editing their colors, checking that everything still works. I can not implement it in this PR.